File: Python/695. Max Area of Island.py
class Solution:
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        dx = [0, 1, -1, 0]
        dy = [1, 0, 0, -1]
        maple = dict()
        def bfs(r, c, x, y):
            q = list()
            q.append([x, y])
            maple[(x, y)] = 1
            cnt = 1
            while len(q) > 0:
                now = q.pop(0)
                for i in range(0,4):
                    tx = now[0] + dx[i]
                    ty = now[1] + dy[i]
                    if tx < 0 or ty < 0:
                        continue
                    if tx >= r or ty >= c:
                        continue
                    if maple.get((tx,ty)) is None and grid[tx][ty] == 1:
                        maple[(tx,ty)] = 1
                        q.append([tx,ty])
                        cnt += 1
            return cnt
                
        r = len(grid)
        c = len(grid[0])
        ans = 0
        for i in range(r):
            for j in range(c):
                if grid[i][j] == 1 and maple.get((i, j)) is None:
                    ans = max(ans,  bfs(r, c, i, j))
                   
        return ans

File: C++/695. Max Area of Island.cpp
class Solution {
public:
    int vis[105][105];
    int bfs(int n,int m,int x,int y,vector<vector<int>>& grid)
    {
        
        queue<pair<int,int> >q;
        int dx[] = {0,1,-1,0};
        int dy[] = {1,0,0,-1};
        q.push({x,y});
        int cnt = 1;
        while(!q.empty())
        {
            pair<int,int> now = q.front();
            q.pop();
            for(int i=0;i<4;i++)
            {
                int tx = now.first+dx[i];
                int ty = now.second+dy[i];
                if(tx<0 || ty<0 || tx>=n || ty>=m)
                    continue;
                if(!vis[tx][ty] && grid[tx][ty])
                {
                    vis[tx][ty] = 1;
                    cnt++;
                    q.push({tx,ty});
                }
                    
            }
        }
        return cnt;
    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int ans = 0;
        memset(vis,0,sizeof(vis));
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(vis[i][j]==0 && grid[i][j])
                {
                    vis[i][j] = 1;
                    ans = max(ans, bfs(n,m,i,j,grid));
                }
            }
        }
        return ans;
    }
};